File: 2022-2023/sourcecode/lab/lab11/VeryNobleAppJava/app/src/main/java/pl/udu/uwr/pum/verynobleappjava/ui/fragments/laureate/LaureateViewModel.java
package pl.udu.uwr.pum.verynobleappjava.ui.fragments.laureate;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import pl.udu.uwr.pum.verynobleappjava.data.laureateresponse.LaureateResponse;
import pl.udu.uwr.pum.verynobleappjava.repository.NobelRepository;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class LaureateViewModel extends ViewModel {
    private final NobelRepository repository = new NobelRepository();
    private final MutableLiveData<LaureateResponse> laureate = new MutableLiveData<>();

    private final String TAG = "LaureateViewModel";

    public void getLaureate(String id){
        Call<LaureateResponse> call = repository.getLaureate(id);

        call.enqueue(new Callback<LaureateResponse>() {
            @Override
            public void onResponse(@NonNull Call<LaureateResponse> call, @NonNull Response<LaureateResponse> response) {
                if (response.isSuccessful()){
                    LaureateResponse laureateResponse = response.body();
                    if (laureateResponse != null)
                        laureate.postValue(laureateResponse);
                }
            }

            @Override
            public void onFailure(@NonNull Call<LaureateResponse> call, @NonNull Throwable t) {
                Log.e(TAG, "error: " + t.getMessage() + "at " + TAG);
            }
        });
    }

    public LiveData<LaureateResponse> getLaureate() {
        return laureate;
    }
}

File: 2022-2023/sourcecode/lab/lab11/VeryNobleAppKotlin/app/src/main/java/pl/udu/uwr/pum/verynobleappkotlin/ui/fragments/laureate/LaureateViewModel.kt
package pl.udu.uwr.pum.verynobleappkotlin.ui.fragments.laureate

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import pl.udu.uwr.pum.verynobleappkotlin.data.laureateresponse.LaureateResponse
import pl.udu.uwr.pum.verynobleappkotlin.repository.NobelRepository
import pl.udu.uwr.pum.verynobleappkotlin.util.Resource
import retrofit2.Response

class LaureateViewModel : ViewModel() {
    private val repository = NobelRepository()
    private val _laureate: MutableLiveData<Resource<LaureateResponse>> = MutableLiveData()

    val laureate: LiveData<Resource<LaureateResponse>>
        get() = _laureate

    fun getLaureate(id: String) = viewModelScope.launch {
        _laureate.postValue(Resource.Loading())
        val response = repository.getLaureate(id)
        _laureate.postValue(handleLaureateResponse(response))
    }

    private fun handleLaureateResponse(response: Response<LaureateResponse>)
            : Resource<LaureateResponse> {
        if (response.isSuccessful)
            response.body()?.let { return Resource.Success(it) }
        return Resource.Error(response.message())
    }
}